xqlint error causes "run test" to silently fail

When running a test with the spec-compliant map constructor syntax, eXide silently fails, with this console error:

```
Error while parsing XQuery: syntax error, found ':'
while expecting ':='
at line 9, column 14:
...: "Sunday",
        "Mo" : "Monday",
        "Tu" : "Tuesday",
 ...
```

Correcting the xqlint parser at https://github.com/wolfgangmm/xqlint/blob/master/lib/XQueryParser.ebnf#L426 should solve this immediate issue, though it's unfortunate that the parsing error isn't reported explicitly to the user in the application.

We should probably test atom-existdb to see if its xqlint fork also shares this issue. https://github.com/eXist-db/atom-existdb/blob/master/package.json#L47 suggests that it's using this fork: https://github.com/eXistSolutions/xqlint/tree/master/lib. I can't actually find any trace of support for the map constructor in this repo...

@ccheraa @plutonik-a While some of the original tests given here do now pass, the issue can still be triggered if you use syntax that xqlint chokes on. For an example, follow the steps in the linked issue [above](https://github.com/eXist-db/eXide/issues/224#ref-issue-619240977): https://github.com/eXist-db/exist/issues/3417. I'll add the test here for convenience. 

**Steps to reproduce**

1. Save the following XQSuite test query to the database as `/db/test.xq`

    ```xquery
    xquery version "3.1";
    
    module namespace t="http://exist-db.org/xquery/test";
    
    declare namespace test="http://exist-db.org/xquery/xqsuite";
    
    declare variable $t:XML := document {
        <test>
           <strange xml:id="foo"/>
        </test>
    };
    
    declare variable $t:XSL := document {
        <xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="3.0">
            <xsl:mode on-no-match="shallow-copy"/>
        </xsl:stylesheet>
    };
    
    declare variable $t:CONTROLLER := ``[xquery version "3.1";
    
    if (ends-with($exist:resource, ".xml")) then
         <dispatch xmlns="http://exist.sourceforge.net/NS/exist">
             <view>
                 <forward servlet="XSLTServlet">
                     <set-attribute name="xslt.stylesheet" value="{$exist:root}{$exist:controller}/test.xsl"/>
                 </forward>
             </view>
             <cache-control cache="no"/>
         </dispatch>
     
    else
        ()
    ]``;
    
    declare
        %test:setUp
    function t:setup() {
        let $testCol := xmldb:create-collection("/db/apps", "test")
        return
            (
                xmldb:store("/db/apps/test", "test.xml", $t:XML),
                xmldb:store("/db/apps/test", "test.xsl", $t:XSL),
                xmldb:store("/db/apps/test", "controller.xql", $t:CONTROLLER),
                sm:chmod(xs:anyURI("/db/apps/test/controller.xql"), "o+x")
            )
    };
    
    declare
        %test:tearDown
    function t:tearDown() {
        xmldb:remove("/db/apps/test")
    };
    
    declare
        %test:assertEquals('foo')
    function t:pass-through-controller() {
        doc("http://localhost:8080/exist/apps/test/test.xml")//@xml:id/string()
    };
    
    declare
        %test:assertEquals('foo')
    function t:directly-transform() {
        let $node-tree := doc("/db/apps/test/test.xml")
        let $stylesheet := doc("/db/apps/test/test.xsl")
        let $parameters := ()
        return
            transform:transform($node-tree, $stylesheet, $parameters)//@xml:id/string()
    };
    ```

2. In eXide, run XQuery > Run as test. The expected test results are not shown. No user-facing error is shown. Examine the Javascript console and notice the error (pasted in below the query).

    ```
    Running xqlint...
    util.js:368 Error while parsing XQuery: lexical analysis failed
    while expecting [Wildcard, EQName, IntegerLiteral, DecimalLiteral, DoubleLiteral, StringLiteral, S, '$', '%', '(', '(#', '(:', '+', '-', '.', '..', '/', '//', '<', '<!--', '<?', '@', '[', 'after', 'allowing', 'ancestor', 'ancestor-or-self', 'and', 'append', 'array', 'as', 'ascending', 'at', 'attribute', 'base-uri', 'before', 'boundary-space', 'break', 'case', 'cast', 'castable', 'catch', 'child', 'collation', 'comment', 'constraint', 'construction', 'context', 'continue', 'copy', 'copy-namespaces', 'count', 'decimal-format', 'declare', 'default', 'delete', 'descendant', 'descendant-or-self', 'descending', 'div', 'document', 'document-node', 'element', 'else', 'empty', 'empty-sequence', 'encoding', 'end', 'eq', 'every', 'except', 'exit', 'external', 'first', 'following', 'following-sibling', 'for', 'ft-option', 'function', 'ge', 'group', 'gt', 'idiv', 'if', 'import', 'in', 'index', 'insert', 'instance', 'integrity', 'intersect', 'into', 'is', 'item', 'json-item', 'last', 'lax', 'le', 'let', 'loop', 'lt', 'map', 'mod', 'modify', 'module', 'namespace', 'namespace-node', 'ne', 'node', 'nodes', 'object', 'only', 'option', 'or', 'order', 'ordered', 'ordering', 'parent', 'preceding', 'preceding-sibling', 'processing-instruction', 'rename', 'replace', 'return', 'returning', 'revalidation', 'satisfies', 'schema', 'schema-attribute', 'schema-element', 'score', 'self', 'sliding', 'some', 'stable', 'start', 'strict', 'switch', 'text', 'to', 'treat', 'try', 'tumbling', 'type', 'typeswitch', 'union', 'unordered', 'updating', 'validate', 'value', 'variable', 'version', 'where', 'while', 'with', 'xquery', '{', '{|']
    at line 19, column 35:
    ...``[xquery version "3.1";
    
    if (ends-with($exist:resource, ".xml")...
    ```

3. To confirm that the test can be run, use the query posted [above](https://github.com/eXist-db/eXide/issues/224#issuecomment-536747255)

    ```xquery
    xquery version "3.1";
    
    import module namespace test="http://exist-db.org/xquery/xqsuite" at "resource:org/exist/xquery/lib/xqsuite/xqsuite.xql";
    
    test:suite(
        inspect:module-functions(xs:anyURI("/db/test.xq"))
    )
    ```

The Javascript console error shows that, as described in the original post, an xqlint error can cause eXide's "Run as test" to silently fail. 

Thus, **part 1** of the fix to this issue is to apply a necessary band aid for cases when xqlint's parser chokes on a query. Specifically, eXide should catch xqlint errors like this and report the xqlint parsing error to the user. Something like this from the console above should be shown:

> Error while parsing XQuery: lexical analysis failed

It should probably be shown where other eXide errors are shown - in red text in the lower right corner of the editor pane.

Judging by the location of the parsing error - at line 19 column 35 of the query - xqlint is tripping on test.xq's valid use of the XQuery 3.1 [String Constructor](https://www.w3.org/TR/xquery-31/#id-string-constructors) syntax, i.e., ` ``[ ]`` `.

Looking through the EBNF in the parser, the version of xqlint being used by eXide doesn't have a definition for the String Constructor. 

Thus, **part 2** of the fix is to update the EBNF to add the String Constructor and update the parser used by eXide. For one such PR, see https://github.com/eXist-db/xqlint/pull/1.

For additional context, Note that the xqlint used by eXide is a very old fork of https://github.com/wcandillon/xqlint. That project has since updated its tooling to use more modern Javascript, but in a way that @wolfgangmm found difficult to update eXide to use. But he did use [a newer fork of the upstream xqlint repository](https://github.com/eXistSolutions/xqlint) for his [atom](https://github.com/eXist-db/atom-existdb/blob/master/package.json#L47-L50) and [vscode plugins](https://github.com/wolfgangmm/existdb-langserver/blob/master/server/package.json#L26). The xqlint fork used for those plugins is now 38 commits behind the original fork. It doesn't have the needed EBNF additions for the string constructor. But this is all to say that one alternative approach to "part 2" here is to switch from the old fork to the current upstream master. The EBNF additions would still be needed (hopefully, they could simply be copied from the XQuery spec linked above), but at least the tooling for building xqlint would be much newer.

--- a/src/xquery-helper.js
+++ b/src/xquery-helper.js
@@ -163,6 +163,10 @@ eXide.edit.XQueryModeHelper = (function () {
 			},
 			contentType: "application/octet-stream",
 			success: function (data) {
+                if (data.result !== 'fail' && doc.error) {
+                    data.result = 'fail';
+                    data.error = doc.error;
+                }
 				var valid = $this.compileError(data, doc);
                 if (onComplete) {
 				    onComplete.call(this, valid);
@@ -206,6 +210,7 @@ eXide.edit.XQueryModeHelper = (function () {
             return;
         }
         $.log("Running xqlint...");
+        doc.error = false;
         var session = doc.getSession();
         var value = doc.getText();    
         var h = new JSONParseTreeHandler(value);
@@ -214,6 +219,7 @@ eXide.edit.XQueryModeHelper = (function () {
             parser.parse_XQuery();
         } catch(e) {
             $.log("Error while parsing XQuery: %s", parser.getErrorMessage(e));
+            doc.error = parser.getErrorMessage(e);
             if(e instanceof parser.ParseException) {
                 h.closeParseTree();
             }
@@ -255,6 +261,7 @@ eXide.edit.XQueryModeHelper = (function () {
             session.setAnnotations(annotations);
         } catch(e) {
             $.log("Error while processing ast: %s", e.message);
+            doc.error = doc.error || e.message;
         }
     };
     

